r/geoarrow/R/vctr.R

```diff
@@ -0,0 +1,154 @@
+
+as_geoarrow_vctr <- function(x, ..., schema = NULL) {
+  if (inherits(x, "geoarrow_vctr") && is.null(schema)) {
+    return(x)
+  }
+
+  stream <- as_geoarrow_array_stream(x, ..., schema = schema)
+  chunks <- nanoarrow::collect_array_stream(stream, validate = FALSE)
+  new_geoarrow_vctr(chunks, stream$get_schema())
+}
+
+new_geoarrow_vctr <- function(chunks, schema, indices = NULL) {
+  offsets <- .Call(geoarrow_c_vctr_chunk_offsets, chunks)
+  if (is.null(indices)) {
+    indices <- seq_len(offsets[length(offsets)])
+  }
+
+  structure(
+    indices,
+    schema = schema,
+    chunks = chunks,
+    offsets = offsets,
+    class = c("geoarrow_vctr", "wk_vctr")
+  )
+}
+
+#' @export
+`[.geoarrow_vctr` <- function(x, i) {
+  attrs <- attributes(x)
+  x <- NextMethod()
+  # Assert slice?
+  attributes(x) <- attrs
+  x
+}
+
+#' @export
+infer_nanoarrow_schema.geoarrow_vctr <- function(x, ...) {
+  attr(x, "schema")
+}
+
+#' @export
+as_geoarrow_array_stream.geoarrow_vctr <- function(x, ..., schema = NULL) {
+  as_nanoarrow_array_stream.geoarrow_vctr(x, ..., schema = NULL)
+}
+
+#' @importFrom nanoarrow as_nanoarrow_array_stream
+#' @export
+as_nanoarrow_array_stream.geoarrow_vctr <- function(x, ..., schema = NULL) {
+  slice <- vctr_as_slice(x)
+  if (is.null(slice)) {
+    stop("Can't resolve non-slice geoarrow_vctr to nanoarrow_array_stream")
+  }
+
+  x_schema <- attr(x, "schema")
+
+  # Zero-size slice can be an array stream with zero batches
+  if (slice[2] == 0) {
+    return(nanoarrow::basic_array_stream(list(), schema = x_schema))
+  }
+
+  # Full slice doesn't need slicing logic
+  offsets <- attr(x, "offsets")
+  batches <- attr(x, "chunks")
+  if (slice[1] == 0 && slice[2] == max(offsets)) {
+    return(
+      nanoarrow::basic_array_stream(
+        batches,
+        schema = x_schema,
+        validate = FALSE
+      )
+    )
+  }
+
+  # Calculate first and last slice information
```

Comment:
Can I suggest 1-indexing for these variables? We need 1-indexed values for slicing offsets and chunks. We need 0-indexed offset `nanoarrow::nanoarrow_array_modify()` calls.

I think 1-indexing simplifies this _slightly_.